documents/development/README.md

```diff
@@ -0,0 +1,539 @@
+- C++ libraries should support C++14.
+  - Exception are `celix::Promises` and `celix::PushStreams` which requires C++17.
+- C++ support for `celix::framework` and `celix::utils` must be header-only. 
+- Unit test code can be written in C++17.
+
+## Library target properties
+
+For C and C++ shared libraries, the following target properties should be set:
+ - `VERSION` should be set to the library version.
+ - `SOVERSION` should be set to the library major version.
+ - `OUTPUT_NAME` should be set to the library name and should contain a `celix_` prefix.
+
+```cmake
+add_library(my_lib SHARED
+    src/my_lib.c)
+set_target_properties(my_lib 
+    PROPERTIES
+        VERSION 1.0.0
+        SOVERSION 1
+        OUTPUT_NAME celix_my_lib)
+```
+
+For C and C++ static libraries, the following target properties should be set:
+ - `POSITION_INDEPENDENT_CODE` should be set to `ON` for static libraries.
+ - `OUTPUT_NAME` should be set to the library name and should contain a `celix_` prefix.
+
+```cmake
+add_library(my_lib STATIC
+    src/my_lib.c)
+set_target_properties(my_lib
+    PROPERTIES
+        POSITION_INDEPENDENT_CODE ON
+        OUTPUT_NAME celix_my_lib)
+```
+
+## Symbol Visibility
+
+- Header-only (INTERFACE) libraries should not configure symbol visibility.
+- Shared and static libraries should configure symbol visibility.
+  - Static library meant to be linked as PRIVATE should hide symbols.
+- Bundles should configure symbol visibility (this is done by default).
+
+### Configuring Symbol Visibility for C/C++ Libraries
+
+For Apache Celix shared libraries, symbol visibility should be configured using the CMake target 
+properties `C_VISIBILITY_PRESET`, `CXX_VISIBILITY_PRESET` and `VISIBILITY_INLINES_HIDDEN` and a generated export 
+header. 
+
+The `C_VISIBILITY_PRESET` and `CXX_VISIBILITY_PRESET` target properties can be used to configure the default visibility 
+of symbols in C and C++ code. The `VISIBILITY_INLINES_HIDDEN` property can be used to configure the visibility of 
+inline functions. The `VISIBILITY_INLINES_HIDDEN` property is only supported for C++ code.
+
+The default visibility should be configured to hidden and symbols should be explicitly exported using the export
+marcos from a generated export header. The export header can be generated using the CMake function 
+`generate_export_header`. Every library should have its own export header. 
+
+For shared libraries, this can be done using the following CMake code:
+
+```cmake
+add_library(my_lib SHARED
+        src/my_lib.c)
+set_target_properties(my_lib PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        #For C++ shared libraries also configure CXX_VISIBILITY_PRESET
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN ON
+        OUTPUT_NAME celix_my_lib)
+target_include_directories(my_lib
+      PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/celix/gen/includes/my_lib>
+        $<INSTALL_INTERFACE:include/celix_my_lib>
```

Comment:
```suggestion
```

I think `INSTALL_INTERFACE` is redundant, since `INCLUDES DESTINATION` in install already sets the include directory for the my_lib target.